_source/_posts/2020-04-09-build-fast-website-csharp-dapper.md

```diff
@@ -0,0 +1,771 @@
+
+Go back to SQL Server Management Studio. In the StackOverflow2010 database, open **Storage**, right-click on **Full-Text Catalogs** and click new **Full-Text Catalog**.
+
+{% img blog/csharp-dapper/13-new-full-text-catalog.png alt:"New full text catalog" width:"600" %}{: .center-image }
+
+In this window, go to *Tables/Views* and click dbo.Posts on the left window and click the move to the right button. Then click **Body, Tags, and Title** for our indexed columns.
+
+{% img blog/csharp-dapper/14-full-text-settings.png alt:"Text settings" width:"800" %}{: .center-image }
+
+Since we are in SQL, let's go ahead and add a few more indexes that will be useful given our business requirements.
+
+Right-click dbo.Posts and click Design.
+
+{% img blog/csharp-dapper/15-design.png alt:"Posts table design" width:"600" %}{: .center-image }
+
+Right-click anywhere and click *Indexes/Keys*.
+
+{% img blog/csharp-dapper/16-click-indexes.png alt:"Indexes/keys" width:"800" %}{: .center-image }
+
+Click **Add** and under the name field call it *IX_Posts_Type_Owner*.
+
+Then click on the Columns field, then click the ellipsis.
+
+{% img blog/csharp-dapper/17-click-columns.png alt:"Columns" width:"800" %}{: .center-image }
+
+We're going to add two columns: *PostTypeId* and *OwnerUserId*. This will let us quickly sort the table into questions or answers and then filter down to a specific user if we want. This is advantageous because we effectively get two performance boosts for the price of one. If we don't include the *OwnerUserId* in a query but do specify the PostTypeId, we can quickly filter to just questions or answers which then makes for fewer rows SQL has to search through for any subsequent filters.
+
+{% img blog/csharp-dapper/18-keys.png alt:"Add columns" width:"800" %}{: .center-image }
+
+Let's add one more index. This time only for *ParentId*. This will let us rapidly filter down to the answers for a given question.
+
+## Update the service
+
+Let's update the query we are using in our AskAQuestion method to the following;
+
+```csharp
+var v2sql =
+    @"SELECT *
+    FROM
+    PostWithUser
+
+    INNER JOIN
+    FREETEXTTABLE(Posts, (Body,Tags,Title), @Question) textSearch
+    ON
+    PostWithUser.Id = textSearch.[Key]
+
+    WHERE
+    [PostTypeId] = 1
+
+    ORDER BY
+    textSearch.Rank DESC
+
+    OFFSET @Skip ROWS FETCH NEXT @Take ROWS ONLY;";
+```
+
+The FREETEXTABLE is a part of the FullText service in SQL and lets us specify a table, columns, and an arbitrary set of text that it then uses to rank the members of the table. We can then sort by that rank and filter to the questions like before.
+
+Let's see how our results perform now:
+
+{% img blog/csharp-dapper/19-better-results.png alt:"Better results" width:"800" %}{: .center-image }
+
+MUCH BETTER! Now the service is running really fast and is providing some great results.
+
+## Recap
```

Comment:
I'd delete this header and move the "Learn More" one from below to here.

Also, this section should have a link to the GitHub repo and a CTA to follow us on Twitter + subscribe to our YouTube channel.

Reply:
No repo for this one but I've added the CTA.